server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -346,7 +350,11 @@ protected WriteReplicaResult<ShardUpsertRequest> processRequestItemsOnReplica(In
 
         RawIndexer rawIndexer;
         Indexer indexer;
-        if (insertColumns.length == 1 && insertColumns[0].column().equals(DocSysColumns.RAW)) {
```

Comment:
Hi @BaurzhanSakhariev , I am looking at your first commit. I was wondering if we can create `RawType` such that these if-elses can be removed. Then it would be like taking the `else` block where `RawIndexer` can be `ValueIndexer` and be under the `Indexer` together with the indexer for the generated column (which is currently handled within `RawIndexer`). I will look further.

Reply:
Basically my question is if we can allow `rawRef.valueType().valueIndexer()` to return a `RawIndexer`.

Reply:
Feels to me that it somehow contradicts with current abstraction. 

`Indexer` is a top level component taking `Item` and combining multiple `ValueIndexer`-s which are simple and can write data to the source. 

`RawIndexer` is basically combining `Indexer`-s, which have their own bunch of `ValueIndexer`s actually writing to the source.

Let's imagine we made rawIndexer a `ValueIndexer`. Then, `Indexer` writes source like 
```
xContentBuilder.field(reference.column().leafName()); // source key, we cannot say "raw" in the source. Need special handling.
 valueIndexer.indexValue(
    reference.valueType().sanitizeValue(value), // there is no single value in _raw, it needs some parsing
    xContentBuilder,
    addField,
    onDynamicColumn,
    synthetics,
    columnConstraints
);
```

Reply:
Thanks, I was trying to see if there is a way to hide `insertColumns` size checks and fuse `RawIndexer` and `Indexer`.